### PR TITLE
Improvements for the catalog importer

### PR DESCRIPTION
**Pull request by wrousseau**
## Hey, I just made a Pull Request!

Those are first improvements for the catalog importer as referenced in #4683.
This includes:
* An ADR for choices I've made
* Providing a type selection autocomplete field for generating a pull request containing a catalog-info.yaml file for a given git repository
* Renaming text as well as variables to make it more generic (component -> entity)
* Adaptation of the tests
* Changesets

Happy to get feedback, here are pictures

Capture d’écran 2021-02-25 à 19 35 09

Capture d’écran 2021-02-25 à 19 35 16

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4704-xuge4ujjax7uxeqypkdaenelo1y0eeq1.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6037f127fcb22ddcac608682

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx [hunk 1]
```diff
@@ -82,10 +84,11 @@ export function generateEntities(
     kind: e.kind!,
     metadata: {
       ...e.metadata,
-      name: componentName,
+      name: entityName,
     },
     spec: {
       ...e.spec,
+      type,
```
Comment (dhenneke) [inline plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx:91]: Beware: The `/analyze-entities` endpoint does create a single `Component` at date. But in the future, I assume it could also create entities such as `API` where you don't want to replace the `type`. Even my choice to override the name of all entities is questionable in that light.

An alternative here could be here to have a form for each entity that could be configured. But that would require a larger restructuring of the form. That makes me think about a form where one could also add new entities and that is just pre-populated with the entities that were discovered by the backend. Such forms could e.g. be generated based on the JSON-schemas of the entities.

File: docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+---
+id: adrs-adr012
+title: ADR012: Catalog importer editable kind and type
+description: Architecture Decision Record (ADR) log on which fields are to be made chosable by the catalog importer on repositories
+---
+
+## Background
```
Comment (dhenneke) [inline docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md:7]: I'm not sure if this is worth an ADR.

File: plugins/catalog-import/src/components/ImportEntityPage.tsx [hunk 3]
```diff
@@ -57,19 +57,19 @@ export const ImportComponentPage = (opts: StepperProviderOpts) => {
 
   return (
     <Page themeId="home">
-      <Header title="Register an existing component" />
+      <Header title="Register an existing entity" />
```
Comment (dhenneke) [inline plugins/catalog-import/src/components/ImportEntityPage.tsx:60]: I'm not sure if entity is the best naming-choice here, but I'm also not sure what to use instead. Is the end-user of a Backstage-based portal aware of the meaning of "entity"?

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4704-xuge4ujjax7uxeqypkdaenelo1y0eeq1.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6037f127fcb22ddcac608682

File: plugins/catalog-import/src/components/ImportEntityPage.tsx (hunk 3 above)
Comment (wrousseau) [reply plugins/catalog-import/src/components/ImportEntityPage.tsx:60]: I feel like the correct generic term that refers to a component, an API etc... is an entity based on Backstage's documentation as well as several places on this page that already referred to entities rather than components

File: docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md (hunk 2 above)
Comment (wrousseau) [reply docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md:7]: I've hesitated as I've indeed seen very few ADRs in the repository. However, a new contributor might read https://backstage.io/docs/architecture-decisions/adrs-overview which then refers to https://engineering.atspotify.com/2020/04/14/when-should-i-write-an-architecture-decision-record/

The blog post suggests I should indeed make an ADR for this change. So I think we either need to do more ADRs, or to update the documentation if we want to produce fewer. What do you think?

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx [hunk 4]
```diff
@@ -217,14 +230,30 @@ export function defaultGenerateStepper(
...
-                      error={Boolean(errors.componentName)}
+                      error={Boolean(errors.entityName)}
+                      required
+                    />
+
+                    <AutocompleteTextField
+                      name="type"
+                      control={control}
+                      errors={errors}
+                      freeSolo={false}
```
[the remark this answers is not in the document]
Comment (wrousseau) [reply plugins/catalog-import/src/components/ImportStepper/defaults.tsx:248]: I thought otherwise when coding it but I think you're right. What about just suggesting the list of "standard" types while allowing the user to freely type something else? Eventually, there might be away for users or admins to configure their own list of supported types?

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx (hunk 1 above)
Comment (wrousseau) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx:91]: Yes, I thought about that. i actually have a good use-case : a monorepo which contains multiple front-ends directories (websites) and multiple microservices directories (services). The endpoint should ideally detect:
* all services
* all websites
* APIs which are provided and consumed by the services (much more complex)

But as long as the endpoint does no such thing, I think it's an improvement to allow the user to pick the type of the "main component" of this repository. Does not prevent him from then adding extra catalog-info.yaml files in the repository to cover complex monorepos.

In summary : I suggest we keep this change for now. What do you think?

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (wrousseau) [reply plugins/catalog-import/src/components/ImportStepper/defaults.tsx:248]: In summary if you agree, I'll remove the freeSolo option

File: plugins/catalog-import/src/components/ImportEntityPage.tsx (hunk 3 above)
Comment (Fox32) [reply plugins/catalog-import/src/components/ImportEntityPage.tsx:60]: I would say entity if fine here on the page and in some other parts. Just the create PR part is designed only for components right now, we should keep the wording there.

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx [hunk 5]
```diff
@@ -71,16 +71,29 @@ export type StepperProvider = {
...
+ *
+ * @see https://backstage.io/docs/features/software-catalog/descriptor-format#spectype-required
+ */
+enum CreatableFromRepositoryEntityType {
+  service = 'service',
+  website = 'website',
+  library = 'library',
+  documentation = 'documentation',
+  other = 'other',
+}
```
Comment (Fox32) [inline plugins/catalog-import/src/components/ImportStepper/defaults.tsx:85]: I think this should be configurable, especially as the Backstage documentation itself thinks that using `other` [is not a good idea](https://backstage.io/docs/features/software-catalog/extending-the-model#adding-a-new-type-of-an-existing-kind).

File: plugins/catalog-import/src/components/ImportEntityPage.tsx (hunk 3 above)
Comment (dhenneke) [reply plugins/catalog-import/src/components/ImportEntityPage.tsx:60]: Btw: The link in the Scaffolder uses "Register Existing Component" to link to the catalog-import:
https://github.com/backstage/backstage/blob/118df55e723cb8cc398cad4705e7f39a350bd0ef/plugins/scaffolder/src/components/ScaffolderPage/ScaffolderPage.tsx#L134-L141

They should probably be aligned to the same wording to keep the "create something new / add something existing" storylines aligned.

File: docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md (hunk 2 above)
Comment (andrewthauer) [reply docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md:7]: Although I love the efforts, this is more of a feature that does not necessarily have an broader technology or architecture impact on the code base, so therefore probably doesn't require or warrant and ADR.

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx (hunk 5 above)
Comment (andrewthauer) [reply plugins/catalog-import/src/components/ImportStepper/defaults.tsx:85]: 100% agree. Having these hard-coded is going to be a non starter for lots of people. @Rugvip mentioned something about maybe added a proper API around exposing the allowed/recommended list for an org maybe in the `app-config`. This could most likely use that approach.

Comment (andrewthauer) [reply plugins/catalog-import/src/components/ImportStepper/defaults.tsx:85]: Also, I think another comment touches on this, but there will be different type sets for different entity kinds (e.g. `API => (openapi, graphql, asyncapi, etc.)`)

File: docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md (hunk 2 above)
Comment (wrousseau) [reply docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md:7]: Alright, @andrewthauer do you feel like we should update the Docs to make it clear which features require an ADR, and which ones are covered enough through the elements you mention?

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx (hunk 5 above)
Comment (wrousseau) [reply plugins/catalog-import/src/components/ImportStepper/defaults.tsx:85]: Something like this? We could have an "allow" approach which whitelists types or a "recommended" approach to suggest types in this form. Or both but maybe a bit complex.

If the configuration includes `allow` for the `Component` kind, the user will only be able to select one of these types and won't be able to freely type. If the configuration includes `recommended` for the `Component` kind, the user will have those recommended but will be able to freely type something else. If both are present, the select will suggest the intersection of both and the user will not be able to write something else.

Capture d’écran 2021-02-27 à 12 19 11

File: docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md (hunk 2 above)
Comment (dtuite) [reply docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md:7]: I would expect an ADR to be created when there is difference of opinion on a matter, debate happens, and then the decision is recorded. In this case it seems like everyone generally agrees and thus the value of adding an ADR is limited.

With that said, personally I don't see a significant cost to adding an ADR so I wouldn't request changes based on this. Others may disagree.

Comment (Rugvip) [reply docs/architecture-decisions/adr012-catalog-importer-editable-kind-and-type.md:7]: Would also say that ADRs are generally how to write code rather than how to implement features. For example what languages to use, code style, libraries, particular patterns to use or avoid, etc. So +1 to not needing an ADR for this change

File: plugins/catalog-import/src/components/ImportStepper/defaults.tsx (hunk 5 above)
Comment (andrewthauer) [reply plugins/catalog-import/src/components/ImportStepper/defaults.tsx:85]: Allow for recommended might be different for each org. I'd say starting with a recommend list makes most sense. I've come up with a few cases where I'm entirely sure how to classify something under the existing recommended types and end up throwing them in `other` for now (with the intention of cleaning these up later. Some examples are repos that contain IaaC, docker images, demos, cloud functions, etc.

Perhaps something like this could provide a baseline + room for extension?

```yaml
catalog:
  kinds:
    Component: # not sure on upper-casing this here, but having this keyed would be nice
       typeRule: 'recommended' # or 'restricted'
       types:
         - service
         - website
         - library
         - documentation
```

Comment (Rugvip) [reply plugins/catalog-import/src/components/ImportStepper/defaults.tsx:85]: Having a bit of trouble thinking about how we make this a more general thing, avoiding things like `Component` entities having a `type`. The `type` isn't something that applies to all entities, and some entities might have their own version, either called `type` or something else. Was thinking that something like JSONSchema is already able to express all we need, but that also gets really verbose.

I'm thinking we can head down two paths, either we (1) standardize on `spec.type` being a known field that entities can use to further distinguish between things, or we (2) create a dead simple and limited schema for expressing possible values of an entity field, something like

```yaml
kinds:
  Component:
    fields:
    - path: 'spec.type'
      values:
      - service
      - website
      - library
      - documentation
```

Personally I'm leaning towards (1) as I think it could provide benefit in a bunch of places, but I'm also immediately worried that the need for a 3rd dimension pops up and we're soon dealing with hyperentities. There's also the case that settling (2) would let us describe things that are perhaps not quite as important to how the frontend renders things, but it still nice to put some bounds on, like `spec.lifecycle`.

**Review by Rugvip — COMMENTED** [review]
Hi, want to check in what the status is on this

For now moving the ADR contents perhaps to the PR body would make sense. Also wondering if you have any thoughts on the discussion around configuring the listed types?
